### PR TITLE
Contract: New contract for caching blockchain

### PR DESCRIPTION
**Pull request by fangyi-zhou**
Abstract: We try to modify contract in order to cache the blockchain for more efficient access. This is first step.

We sync all contracts initially and subscribe events from both Master contract and Energy contract. We wait for events to fire so that we don't need to sync the blockchain again every time.

The amount of work of doing this is non-trivial. Let's work on a cache-blockchain branch, but before that, I would like to use PRs to make sure that small changes are reviewed first before we continue.

The first PR is on modification of the contract. Upon that we can reorganise codes in dapp and meter client accordingly. Please review.

### CODE REVIEW

**Comment by fangyi-zhou** [comment]
Note: this does not go to master yet

File: truffle/contracts/Energy.sol [hunk 1]
```diff
@@ -6,32 +6,39 @@ contract Energy {
     uint public unitPrice;
     uint public offeredAmount;
     mapping (address => uint) public remainingEnergy;
+    uint public boughtAmount;  // sum of remainingEnergy;
     address public seller;
-    address public master_addr;
+    bool public noUpdate;
```
Comment (vincepandolfo) [inline truffle/contracts/Energy.sol]: where is `noUpdate` used?

File: truffle/contracts/Master.sol [hunk 2]
```diff
@@ -3,37 +3,23 @@ pragma solidity ^0.4.10;
...
-        address contract_addr;
-        address seller;
-        bool deregistered;
-    }
 
-    ContractEntity[] public contracts;
-    // lookup index of contract in the list
-    mapping (address => uint) lookupIdxByContract;
+    address[] public contracts;
+    event Update(uint index);
```
Comment (vincepandolfo) [inline truffle/contracts/Master.sol]: Is there a reason why this is `Update` and not `update like in the other contract?

File: truffle/contracts/Energy.sol (hunk 1 above)
Comment (fangyi-zhou) [reply truffle/contracts/Energy.sol]: during blockchain sync, a client would like to know whether a contract needs to be subscribed for event updates

File: truffle/contracts/Master.sol (hunk 2 above)
Comment (fangyi-zhou) [reply truffle/contracts/Master.sol]: i need to change other ones, CapCamelCase is recommended for events

Comment (vincepandolfo) [reply truffle/contracts/Master.sol]: will approve after that change

**Comment by fangyi-zhou** [comment]
We need to write tests for those events. Hopefully there won't be a contract fix in the near future.

File: truffle/contracts/Energy.sol [hunk 3]
```diff
@@ -6,32 +6,39 @@ contract Energy {
...
+    event update(uint blockNumber);
 
     function Energy(address seller_, uint unitPrice_, uint offeredAmount_) public {
         unitPrice = unitPrice_;
         offeredAmount = offeredAmount_;
         seller = seller_;
-        // Sender is master contract, this is called in sell function
-        master_addr = msg.sender;
+        boughtAmount = 0;
+        noUpdate = false;
```
Comment (danzvara) [inline truffle/contracts/Energy.sol]: nitpicking, but why noUpdate = false instead of update = true ? :D

**Comment by fangyi-zhou** [comment]
The better question to ask is whether we can query whether an event has occurred or not so that we don’t need the bool.
I was coding this in a tube without Internet so I wasn’t able to check.

**Comment by fangyi-zhou** [comment]
If we really care, setting it to false like I did saves us 1 operation (initialisation to true)

**Comment by danzvara** [comment]
Idk, i'd say we should be able to subscribe so we can listen to event's state.

**Comment by fangyi-zhou** [comment]
Yes that is expected, but can we find out whether an event has occurred or not?

There was a previous version with 2 events, Update and FinalUpdate. The client subscribe to both. If FinalUpdate event has fired, the client should unsubscribe.

I then recognise that if we cannot find out whether an event has been fired, we cannot tell, so I made it a bool and removed the event FinalUpdate.

**Comment by danzvara** [comment]
We should be able to track FinalUpdate too - like this. If it's called, unsubscribe on callback.
```
event.watch(function(error, result){
    // result will contain various information including the argumets given to the Deposit call.
    if (!error)
        console.log(result);
});
 ```

**Comment by fangyi-zhou** [comment]
I think you didn’t get my point.

The reason I removed FinalUpdate is that we don’t know whether it has occurred in the past during initial sync. This is replaced with a bool because it’s basically a latch.

**Comment by danzvara** [comment]
Eh, I see. Isn't it possible to find it out from chain's log? Anyway I'm out of depth here, so LGTM then.

**Comment by fangyi-zhou** [comment]
I think you probably can, but it’s non-trivial.

File: truffle/contracts/Energy.sol [hunk 4]
```diff
@@ -6,32 +6,39 @@ contract Energy {
     uint public unitPrice;
     uint public offeredAmount;
     mapping (address => uint) public remainingEnergy;
```
Comment (danzvara) [inline truffle/contracts/Energy.sol]: energyPerBuyer?

Comment (danzvara) [reply truffle/contracts/Energy.sol]: remainingEnergyPerBuyer?

File: truffle/contracts/Energy.sol [hunk 5]
```diff
@@ -6,32 +6,39 @@ contract Energy {
...
+    uint public boughtAmount;  // sum of remainingEnergy;
```
Comment (danzvara) [inline truffle/contracts/Energy.sol]: soldAmount - how much we sold and is left to consume (we offer offered amount and keep track of sold amount)

File: truffle/contracts/Master.sol (hunk 2 above)
Comment (danzvara) [reply truffle/contracts/Master.sol]: /noUpdate/empty/s ? you want to indicate contract can be discarded, noUpdate doesn't say that

Comment (fangyi-zhou) [reply truffle/contracts/Master.sol]: good point

File: truffle/contracts/Energy.sol (hunk 4 above)
Comment (fangyi-zhou) [reply truffle/contracts/Energy.sol]: buyerEnergyBalances?

File: truffle/contracts/Energy.sol (hunk 5 above)
Comment (fangyi-zhou) [reply truffle/contracts/Energy.sol]: will fix

File: truffle/contracts/Energy.sol (hunk 4 above)
Comment (danzvara) [reply truffle/contracts/Energy.sol]: balancePerBuyer?

Comment (fangyi-zhou) [reply truffle/contracts/Energy.sol]: ok

**Comment by danzvara** [comment]
Otherwise LGTM, here's your seal. http://diving.ie/site/wp-content/uploads/2012/06/seal10.jpg

**Pull request merged**